Sources/CombineX/ObserableObject.swift

```diff
@@ -35,18 +37,51 @@ public protocol ObservableObject : AnyObject {
     var objectWillChange: Self.ObjectWillChangePublisher { get }
 }
 
-extension ObservableObject where Self.ObjectWillChangePublisher == ObservableObjectPublisher {
+private protocol PublishedProtocol {
+    var objectWillChange: ObservableObjectPublisher? { get set }
+}
+extension Published: PublishedProtocol {}
 
+private let publishedPropertiesCache = Cache<UnsafeRawPointer, [PropertyInfo]>()
+private let globalObjectWillChangeCache = WeakCache<AnyObject, ObservableObjectPublisher>()
+
+extension ObservableObject where Self.ObjectWillChangePublisher == ObservableObjectPublisher {
+    
+    private static func publishedProperties() throws -> [PropertyInfo] {
+        let key = unsafeBitCast(self, to: UnsafeRawPointer.self)
+        if let prop = publishedPropertiesCache[key] {
+            return prop
+        }
+        let info = try typeInfo(of: self)
+        let prop = info.properties.filter { $0.type is PublishedProtocol.Type }
+        publishedPropertiesCache[key] = prop
+        return prop
+    }
+    
+    private func setObservableObjectPublisher(_ objectWillChange: ObservableObjectPublisher, for publishedProperty: PropertyInfo) throws {
+        // TODO: mutate in place
+        var published = try publishedProperty.get(from: self) as! PublishedProtocol
+        published.objectWillChange = objectWillChange
+        try withUnsafePointer(to: self) { ptr in
+            let mptr = UnsafeMutablePointer(mutating: ptr)
+            try publishedProperty.set(value: published, on: &(mptr.pointee))
+        }
+    }
+    
     /// A publisher that emits before the object has changed.
     public var objectWillChange: ObservableObjectPublisher {
-        /*
-         TODO: How to create a default publisher?
-         
-            1: All @Published properties hold this object and send changes.
-            2: This object gets all @Published properties and subscribes to them.
-         https://github.com/apple/swift-evolution/blob/master/proposals/0258-property-wrappers.md#referencing-the-enclosing-self-in-a-wrapper-type
-         */
-        Global.RequiresImplementation()
+        if let pub = globalObjectWillChangeCache[self] {
+            return pub
+        }
+        do {
+            let publishedProperties = try type(of: self).publishedProperties()
+            let pub = ObservableObjectPublisher()
+            try publishedProperties.forEach { try setObservableObjectPublisher(pub, for: $0) }
+            globalObjectWillChangeCache[self] = pub
+            return pub
+        } catch {
+            fatalError(error.localizedDescription)
+        }
```

Comment:
The cache logic is wrong. `if nil { make() }` should be locked, otherwise `make()` might be executed more than once.